In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\rbrand\12-28-2024 Instacart Basket Analysis'

In [3]:
#3.Importing orders_products_prior.csv data set from original data.
df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)

In [4]:
#Importing orders_wrangled.csv data set from prepared data.
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [8]:
#check the output
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [6]:
#check the output
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  order_frequency_days  
0                        8                   NaN  
1                        7                  15.0  
2                       12                  21.0  
3                        7                  29.0  
4                       15                  28.0  
...                    ...                   ...  
3421078                 18                  29.0  
3421079                 10                  30.0  
3421080                 12                  18.0  
3421081                 12                   7.0  
3421082                 14                  30.0  

[3421083 rows x 7 columns]

In [9]:
df_ords_prior.shape

(32434489, 4)

In [10]:
df_ords.shape

(3421083, 7)

In [15]:
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [12]:
df_merged_large.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
0                  8                   NaN         196                  1   
1                  8                   NaN       14084                  2   
2                  8                   NaN       12427                  3   
3                  8                   NaN       26088                  4   
4                  8                   NaN       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [13]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [14]:
# Export data to pkl

df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))

In [ ]:
#Task 4.6---------------------------------------------------------------------------------------------

In [ ]:
#1. Merged
#2. orders_products_combined.pkl exported

In [25]:
#3.Importing orders_products_combined.pkl data set from prepared data.
file_path = os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl')
orders_products_combined = pd.read_pickle(file_path)

In [27]:
orders_products_combined.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
0                  8                   NaN         196                  1   
1                  8                   NaN       14084                  2   
2                  8                   NaN       12427                  3   
3                  8                   NaN       26088                  4   
4                  8                   NaN       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [28]:
#4. CHecking the shape of the imported Data frame
print("Shape of the imported DataFrame:", orders_products_combined.shape)

Shape of the imported DataFrame: (32434489, 11)


In [32]:
#5.#Importing products_checked.csv data set from prepared data.
df_products_checked = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col=False)

In [34]:
df_products_checked.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [43]:
#5.Using 'product_id' as common key between the two dataframes.
merged_data = pd.merge(orders_products_combined, df_products_checked, on='product_id', how='inner', indicator='merge_flag')

In [44]:
#5.Checking first few rows of the merged DataFrame
print(merged_data.head())

   Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
0                  8                   NaN         196                  1   
1                  7                  15.0         196                  1   
2                 12                  21.0         196                  1   
3                  7                  29.0         196                  1   
4                 15                  28.0         196                  1   

   reordered _merge  Unnamed: 0_y product_name  aisle_id  department_id  \
0          0   both          

In [45]:
#5.Checking the shape of the merged DataFrame
print("Shape of the merged DataFrame:", merged_data.shape)

Shape of the merged DataFrame: (32404859, 17)


In [54]:
#6. Checking the merge flag frequency.
merge_flag_counts = merged_data['merge_flag'].value_counts()

In [55]:
#6. Checking results("Merge Flag Value Counts:")
print(merge_flag_counts)

Merge Flag Value Counts:
merge_flag
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64


In the merge operation, the 'merge_flag' column was utilized to indicate the origin of each row in the merged DataFrame. The value counts show that there are 32,404,859 rows where the key ('product_id') exists in both dataframes ('both'). There are no rows exclusive to either the left or right dataframe ('left_only' and 'right_only' both have counts of 0). This result suggests a full match between the two datasets based on the specified key, reinforcing the successful integration of product data into the combined orders and products DataFrame.

In [57]:
#7.Exporting the merged DataFrame as a Pickle file
merged_data.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))